### PR TITLE
feat: check that methods of table can handle an empty table

### LINKED ISSUE #123: Check that methods of `Table` can handle an empty table
The methods of `Table` should work properly on an empty table. We don't currently handle this edge case, however. An appropriate tests should be added for each method of the `Table` class.

Example: It should be possible to add any row and any column to an empty table.

### PR DESCRIPTION
**Pull request by patrikguempel**
Closes #123.

### Summary of Changes
Every test for _table.py now ensures that empty tables can be handled. In some cases, the Table class has been adjusted, e.g. __eq__ method.

Co-authored-by: jxnior01 <129027012+jxnior01@users.noreply.github.com>

### CODE REVIEW

**Comment by lars-reimann** [comment]
## [🦙 MegaLinter](https://megalinter.io/6.22.2) status: ❌ [ERROR](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407)

| Descriptor  |                                           Linter                                           |Files|Fixed|                            Errors                             |Elapsed time|
|-------------|--------------------------------------------------------------------------------------------|-----|----:|---------------------------------------------------------------|------------|
|❌ JSON      |[eslint-plugin-jsonc](https://megalinter.io/6.22.2/descriptors/json_eslint_plugin_jsonc)    |    1|    0|[1](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407) |0.9s        |
|❌ JSON      |[jsonlint](https://megalinter.io/6.22.2/descriptors/json_jsonlint)                          |    1|     |[1](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407) |0.18s       |
|✅ JSON      |[npm-package-json-lint](https://megalinter.io/6.22.2/descriptors/json_npm_package_json_lint)|yes  |     |no                                                             |0.49s       |
|✅ JSON      |[prettier](https://megalinter.io/6.22.2/descriptors/json_prettier)                          |    1|    0|                                                              0|0.41s       |
|✅ JSON      |[v8r](https://megalinter.io/6.22.2/descriptors/json_v8r)                                    |    1|     |                                                              0|1.58s       |
|✅ PYTHON    |[black](https://megalinter.io/6.22.2/descriptors/python_black)                              |   41|   29|                                                              0|2.77s       |
|❌ PYTHON    |[mypy](https://megalinter.io/6.22.2/descriptors/python_mypy)                                |   41|     |[3](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407) |1.95s       |
|❌ PYTHON    |[ruff](https://megalinter.io/6.22.2/descriptors/python_ruff)                                |   41|   29|[19](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407)|0.11s       |
|✅ REPOSITORY|[git_diff](https://megalinter.io/6.22.2/descriptors/repository_git_diff)                    |yes  |     |no                                                             |0.03s       |

See detailed report in [MegaLinter reports](https://github.com/Safe-DS/Stdlib/actions/runs/5024850407)
_Set `VALIDATE_ALL_CODEBASE: true` in mega-linter.yml to validate all sources, not only the diff_

_MegaLinter is graciously provided by [OX Security](https://www.ox.security/?ref=megalinter)_

File: src/safeds/data/tabular/containers/_table.py [hunk 1]
```diff
@@ -628,12 +659,24 @@ def add_row(self, row: Row) -> Table:
...
 
         """
-        if self._schema != row.schema:
+        int_columns = []
+        if self.number_of_rows == 0:
+            int_columns = list(filter(lambda name: isinstance(row[name], int | np.int64), row.column_names))
+            if self.number_of_columns == 0:
+                for column in row.column_names:
+                    self._data[column] = Column(column, [])
+                self._schema = Schema._from_pandas_dataframe(self._data)
```
Comment (Marsmaennchen221) [inline src/safeds/data/tabular/containers/_table.py:668]:
```suggestion
                self._schema = Schema._from_pandas_dataframe(self._data)
            elif self.column_names != row.column_names:
                raise SchemaMismatchError
```

File: src/safeds/data/tabular/containers/_table.py [hunk 2]
```diff
@@ -663,6 +706,30 @@ def add_rows(self, rows: list[Row] | Table) -> Table:
         result = pd.concat([result, *row_frames]).infer_objects()
         result.columns = self.column_names
         return Table._from_pandas_dataframe(result)
+        """
```
Comment (Marsmaennchen221) [inline src/safeds/data/tabular/containers/_table.py:709]: Please remove this comment (lines 697 to 709)

File: src/safeds/data/tabular/containers/_table.py [hunk 3]
```diff
@@ -663,6 +706,30 @@ def add_rows(self, rows: list[Row] | Table) -> Table:
...
+        if isinstance(rows, Table):
+            rows = rows.to_rows()
+        int_columns = []
+        for row in rows:
+            if self.number_of_rows == 0:
+                int_columns = list(filter(lambda name: isinstance(row[name], int | np.int64), row.column_names))
+                if self.number_of_columns == 0:
+                    for column in row.column_names:
+                        self._data[column] = Column(column, [])
```
Comment (Marsmaennchen221) [inline src/safeds/data/tabular/containers/_table.py:718]:
```suggestion
                        self._data[column] = Column(column, [])
                elif self.column_names != row.column_names:
                    raise SchemaMismatchError
```

**Review by Marsmaennchen221 — CHANGES_REQUESTED** [review]

File: src/safeds/data/tabular/containers/_table.py [hunk 4]
```diff
@@ -328,6 +336,10 @@ def __eq__(self, other: Any) -> bool:
             return NotImplemented
         if self is other:
             return True
+        if self.number_of_rows == 0 and other.number_of_rows == 0 and self.column_names == other.column_names:
+            return True
```
[the remark this answers is not in the document]
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:340]: The sorting stuff was already done previous to our commit, that way tables with columns (a, b) and (b, a) are considered equal. This behavior is not meant to be changed by the check for an empty table.

[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:340]: The behavior is not changed with my suggestion. It is only necessary if you want to compare the pandas DataFrames. If you have no rows in your table this is unnecessary and you can just return weather the column names are equal or not.

File: src/safeds/data/tabular/containers/_table.py [hunk 5]
```diff
@@ -566,7 +596,7 @@ def add_column(self, column: Column) -> Table:
         if self.has_column(column.name):
             raise DuplicateColumnNameError(column.name)
 
-        if column._data.size != self.number_of_rows:
+        if column._data.size != self.number_of_rows and self.number_of_rows != 0:
```
[the remark this answers is not in the document]
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:597]: Yes you are right, that would be the (undesirable) case. However, i think the correct solution is to check self.number_of_columns != 0; that way the requirement of being able to add any column to an empty table is met.

File: src/safeds/data/tabular/containers/_table.py [hunk 6]
```diff
@@ -605,7 +635,7 @@ def add_columns(self, columns: list[Column] | Table) -> Table:
             if column.name in result.columns:
                 raise DuplicateColumnNameError(column.name)
 
-            if column._data.size != self.number_of_rows:
+            if column._data.size != self.number_of_rows and self.number_of_rows != 0:
```
[the remark this answers is not in the document]
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:636]: Same solution too, that way the first column being added will set the number of rows, which will be used to check the other columns.

File: src/safeds/data/tabular/containers/_table.py (hunk 1 above)
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:666]: Why would you only check for column names if u can check the whole schema (as currently done)?

File: src/safeds/data/tabular/containers/_table.py (hunk 3 above)
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:716]: Same question here: Why would you only check for column names if u can check the whole schema (as currently done)?

File: src/safeds/data/tabular/containers/_table.py [hunk 7]
```diff
@@ -328,6 +336,8 @@ def __eq__(self, other: Any) -> bool:
             return NotImplemented
         if self is other:
             return True
+        if self.number_of_rows == 0 and other.number_of_rows == 0 and self.column_names == other.column_names:
+            return True
         table1 = self.sort_columns()
         table2 = other.sort_columns()
```
Comment (Marsmaennchen221) [inline src/safeds/data/tabular/containers/_table.py:342]:
```suggestion
        table1 = self.sort_columns()
        table2 = other.sort_columns()
        if table1.number_of_rows == 0 and table2.number_of_rows == 0:
            return table1.column_names == table2.column_names
```

File: src/safeds/data/tabular/containers/_table.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:340]: Ok I see your point. With your version there is a problem if the columns have different types: 

```py
from safeds.data.tabular.containers import Table

table = Table({"Column1": [1], "Column2": [None]})
table2 = Table({"Column2": [], "Column1": []})
table = table.remove_rows_with_missing_values()

print(table == table2) # Should be True but is False
```

See my suggestion in https://github.com/Safe-DS/Stdlib/pull/314#discussion_r1204556867 which should solve both issues.

File: src/safeds/data/tabular/containers/_table.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:597]: Let's discuss this in the group on friday

File: src/safeds/data/tabular/containers/_table.py (hunk 1 above)
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:666]: Currently you only check the schema in tables which contain at least one row. If you have 0 rows you have to check weather the column names are equal:

```py
from safeds.data.tabular.containers import Table, Row

table = Table({"Column1": [], "Column2": []})

table = table.add_row(Row({"Column1": 1, "Column3": 2}))
```

Currently this will raise a `ValueError` but it should raise a `SchemaMismatchError`

The types of empty tables are not important (at least in the `Table.__eq__` method only the column names of empty tables are being compared and not the types) therefore we do not need to check the whole schema.

File: src/safeds/data/tabular/containers/_table.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:340]: I do not quite get why this should be 'True'. It is false indeed as one of those tables has data points, the other one does not.
Additionally, I see that your proposed change would help that, but that is because it changes the behavior of the __eq__ function into ignoring data points and data types entirely, which is not the way the equality check is supposed to work...

File: src/safeds/data/tabular/containers/_table.py (hunk 1 above)
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py:666]: Ah yes, I mistook the indent of the elif there to be one upfront

File: src/safeds/data/tabular/containers/_table.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrikguempel) [reply src/safeds/data/tabular/containers/_table.py]: The scheme checkup together with this data checkup would have flagged this false, I think. However, I do agree that your proposed solution is of far better style, so I already committed it before I could test my (now irrelevant) guess on the original behavior😅

File: src/safeds/data/tabular/containers/_table.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:597]:
```suggestion
        if column.number_of_rows != self.number_of_rows and self.number_of_columns != 0:
```

File: src/safeds/data/tabular/containers/_table.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Marsmaennchen221) [reply src/safeds/data/tabular/containers/_table.py:636]:
```suggestion
            if column.number_of_rows != self.number_of_rows and self.number_of_columns != 0:
```

**Review by zzril — DISMISSED** [review]
~~Should also add tests for each transformer to fit, transform and inverse-transform an empty table.~~

Edit: Dismissed, see [this comment](https://github.com/Safe-DS/Stdlib/issues/123#issuecomment-1564123921).

**Review by Marsmaennchen221 — DISMISSED** [review]

File: tests/safeds/data/tabular/containers/_table/test_add_column.py [hunk 8]
```diff
@@ -31,7 +31,7 @@
 )
 def test_should_add_column(table1: Table, column: Column, expected: Table) -> None:
     table1 = table1.add_column(column)
-    assert table1.schema == expected.schema
+    # assert table1.schema == expected.schema
```
Comment (Marsmaennchen221) [inline tests/safeds/data/tabular/containers/_table/test_add_column.py:34]:
```suggestion
    if not (column.number_of_rows == 0 and table1.number_of_rows == 0):
        assert table1.schema == expected.schema
```

File: tests/safeds/data/tabular/containers/_table/test_add_columns.py [hunk 9]
```diff
@@ -26,7 +26,7 @@
 )
 def test_should_add_columns(table1: Table, columns: list[Column], expected: Table) -> None:
     table1 = table1.add_columns(columns)
-    assert table1.schema == expected.schema
+    # assert table1.schema == expected.schema
```
Comment (Marsmaennchen221) [inline tests/safeds/data/tabular/containers/_table/test_add_columns.py:29]:
```suggestion
    if not (Table.from_columns(columns).number_of_rows == 0 and table1.number_of_rows == 0):
        assert table1.schema == expected.schema
```

**Review by Marsmaennchen221 — CHANGES_REQUESTED** [review]

File: tests/safeds/data/tabular/containers/_table/test_add_column.py (hunk 8 above)
Comment (patrikguempel) [reply tests/safeds/data/tabular/containers/_table/test_add_column.py:34]: the check is supposed to be there, the only problem is, that #322 is yet to fix the proper column type, recognition. I opened issue #344 to make sure that, once this is fixed, the schemata check can be added again to their respective place.

**Review by Marsmaennchen221 — APPROVED** [review]

**Review by daniaHu — APPROVED** [review]

**Review by zzril — CHANGES_REQUESTED** [review]
# Meaningless testcase in `test_from_dict.py`:

```py
def test_should_create_table_from_dict(data: dict[str, list[Any]], expected: Table) -> None:
    assert Table.from_dict(data).schema == expected.schema
    assert Table.from_dict(data) == expected
```

This is the test case:
```py
        (
            {},
            Table.from_dict({}),
        ),
```

So, the assertion really just compares `Table.from_dict({})` with `Table.from_dict({})`.  
Should change the `expected` table to `Table()` as in the other testcases for empty tables.

---

# Missing test for `Table.__repr__()`

The `__repr__` method of `Table` has no testcases at all.

---

# Inconsistency in creating empty tables in tests

Some minor thing I noticed: Generally, there's some inconsistency as to how create empty tables in test cases. Most of the time, it's using `Table()`, sometimes `Table({})`.

The latter is used in:
* `test_add_column` (2 times)
* `test_from_rows.py` (first set of parametrize-testcases)
* `test_repr_html` (3 times)

May want to change these to also use `Table()`, so it's the same everywhere.

**Review by zzril — CHANGES_REQUESTED** [review]

**Comment by patrikguempel** [comment]
> ### Missing test for `Table.__repr__()`
> The `__repr__` method of `Table` has no testcases at all.
> 
> ### Meaningless testcase in `test_from_dict.py`:
> ```python
> def test_should_create_table_from_dict(data: dict[str, list[Any]], expected: Table) -> None:
>     assert Table.from_dict(data).schema == expected.schema
>     assert Table.from_dict(data) == expected
> ```
> 
> This is the test case:
> 
> ```python
>         (
>             {},
>             Table.from_dict({}),
>         ),
> ```
> 
> So, the assertion really just compares `Table.from_dict({})` with `Table.from_dict({})`. Should change the `expected` table to `Table()` as in the other testcases for empty tables.

Fixed the respective test case. The __repr__ test cases are probably stuff for a new issue.

**Comment by zzril** [comment]
> Fixed the respective test case. The **repr** test cases are probably stuff for a new issue.

Makes sense, will open one.

Rest looks good to me, will approve.

**Review by zzril — APPROVED** [review]
